File: advent2023/day01/src/main.rs
use std::cmp::min;

fn first_digit(s: &str) -> Option<u32> {
    for c in s.chars() {
        if c.is_digit(10) {
            return Some(c.to_digit(10).unwrap());
        }
    }
    return None;
}

fn last_digit(s: &str) -> Option<u32> {
    for c in s.chars().rev() {
        if c.is_digit(10) {
            return Some(c.to_digit(10).unwrap());
        }
    }
    return None;
}

const NUMBER_STR: [&str; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

fn first_digit_2(s: &str) -> Option<u32> {
    let chars = s.chars().collect::<Vec<char>>();
    for i in 0..s.len() {
        let c = chars.get(i).unwrap();
        if c.is_digit(10) {
            return Some(c.to_digit(10).unwrap());
        }
        let sub = &s[i..min(i + 5, s.len())];
        for (j, &n) in NUMBER_STR.iter().enumerate() {
            if sub.starts_with(n) {
                return Some((j + 1) as u32);
            }
        }
    }
    None
}

fn last_digit_2(s: &str) -> Option<u32> {
    let chars = s.chars().collect::<Vec<char>>();
    for i in (0..s.len()).rev() {
        let c = chars.get(i).unwrap();
        if c.is_digit(10) {
            return Some(c.to_digit(10).unwrap());
        }
        let sub = &s[i..min(i + 5, s.len())];
        for (j, &n) in NUMBER_STR.iter().enumerate() {
            if sub.starts_with(n) {
                return Some((j + 1) as u32);
            }
        }
    }
    None
}

fn main() {
    let input = std::fs::read_to_string("data/input.txt").unwrap();
    let lines = input.lines();
    let mut total = 0;
    for l in lines.clone() {
        if let (Some(x), Some(y)) = (first_digit(l), last_digit(l)) {
            total += 10 * x + y;
        }
    }
    println!("{}", total);
    let mut total = 0;
    for l in lines {
        if let (Some(x), Some(y)) = (first_digit_2(l), last_digit_2(l)) {
            total += 10 * x + y;
        }
    }
    println!("{}", total);
}

File: advent2023/day01/app/Main.hs
module Main (main) where

import Data.Char (digitToInt, isDigit)
import Data.List (elemIndex)
import Data.Maybe (fromMaybe, maybeToList)

firstDigit :: String -> Int
firstDigit s = digitToInt $ head $ filter isDigit s

isPrefixOf :: String -> String -> Bool
isPrefixOf p s = (take (length p) s) == p

digitStrings :: [String]
digitStrings = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

toDigits :: [Char] -> [Int]
toDigits "" = []
toDigits xs@(s:ss) = 
  (if isDigit s 
    then [digitToInt s]
    else parsed) ++ toDigits ss
  where
    isPrefix = map (\p -> isPrefixOf p xs) digitStrings
    parsed = maybeToList $ (+1) <$> elemIndex True isPrefix

main :: IO ()
main = do
  input <- readFile "data/input.txt"
  let xs = lines input
  let fstDigits = map firstDigit xs
  let lstDigits = map (firstDigit . reverse) xs
  let vals = zipWith (\x y -> read $ concat [show x, show y]) fstDigits lstDigits :: [Int]
  print $ sum vals
  let digits = map toDigits xs
  let vals = map (\x -> read $ concat $ map show [head x, last x]) digits :: [Int]
  print $ sum vals
